internal/yggdrasil/yggdrasil.go

```diff
@@ -299,73 +299,59 @@ func (h *Handler) PostDataMessageForDevice(ctx context.Context, params yggdrasil
 		content := models.RegistrationResponse{}
 
 		dvc, err := h.deviceRepository.Read(ctx, deviceID, h.initialNamespace)
-		if err == nil {
-			isInit := false
-			if dvc.ObjectMeta.Labels[v1alpha1.EdgeDeviceSignedRequestLabelName] == v1alpha1.EdgeDeviceSignedRequestLabelValue {
-				isInit = true
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				h.metrics.IncEdgeDeviceFailedRegistration()
+				return operations.NewPostDataMessageForDeviceInternalServerError()
 			}
+			return operations.NewPostDataMessageForDeviceNotFound()
+		}
 
-			if !isInit && !IsOwnDevice(ctx, deviceID) {
-				authKeyVal, _ := ctx.Value(AuthzKey).(string)
-				logger.V(0).Info("Device tries to re-register with an invalid certificate", "certcn", authKeyVal)
-				// At this moment, the registration certificate it's no longer valid,
-				// because the CR is already created, and need to be a device
-				// certificate.
-				return operations.NewPostDataMessageForDeviceForbidden()
-			}
+		if dvc == nil {
+			h.metrics.IncEdgeDeviceFailedRegistration()
+			return operations.NewPostDataMessageForDeviceInternalServerError()
+		}
 
-			// @TODO remove this IF when MTLS is finished
-			if registrationInfo.CertificateRequest != "" {
-				cert, err := h.mtlsConfig.SignCSR(registrationInfo.CertificateRequest, deviceID)
-				if err != nil {
-					return operations.NewPostDataMessageForDeviceBadRequest()
-				}
-				content.Certificate = string(cert)
-			}
-			res.Content = content
-
-			if isInit {
-				newdevice := dvc.DeepCopy()
-				delete(newdevice.Labels, v1alpha1.EdgeDeviceSignedRequestLabelName)
-				err = h.deviceRepository.Patch(ctx, dvc, newdevice)
-				if err != nil {
-					logger.Error(err, "cannot update edgedevice labels")
-				}
-			}
-			return operations.NewPostDataMessageForDeviceOK().WithPayload(&res)
+		isInit := false
+		if dvc.ObjectMeta.Labels[v1alpha1.EdgeDeviceSignedRequestLabelName] == v1alpha1.EdgeDeviceSignedRequestLabelValue {
+			isInit = true
 		}
 
-		if !errors.IsNotFound(err) {
-			return operations.NewPostDataMessageForDeviceInternalServerError()
+		// the first time that tries to register should be able to use register certificate.
+		if !isInit && !IsOwnDevice(ctx, deviceID) {
```

Comment:
Is it possible for this condition to be used when certificate expires and the device is waiting to registration to be approved?

Reply:
yes, that logic happens on `mtls.VerifyRequest`

Reply:
do I understand correctly that if registration cert has expired and autoapprove process is not enabled the device won't be able to register?